fix(android): handle file: URIs without // after scheme

sgtcoolguy opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-27191

**Optional Description:**
Fix extracted from work on #10998

This specifically fixes support for `file:` URIs that are of the form:
- `file:/path/to/file`
- `file:relative/path`

Before it only supported file URIs that were of the form:
- `file:///path/to/file`

sgtcoolguy committed:
fix(android): handle file: URIs without // after scheme

Fixes TIMOB-27191

build commented:
<!--
  2 failure:  :microscope: Ther..., Tests have failed...
  0 warning: 
  3 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/sgtcoolguy/titanium_mobile/blob/TIMOB-27191/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-11098/1/artifact/dist/mobilesdk-8.2.0.v20190731104707-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 1 tests have failed
There are 1 tests failing and 470 skipped out of 3728 total tests.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>ios.Titanium.Database</td><td>#executeAllAsync() does not allow DB to be GC'd</td><td>120.007</td><td><pre>Error: timeout of 120000ms exceeded
file:///Users/build/Library/Developer/CoreSimulator/Devices/60D909FD-31E6-42C0-959A-8B0B37F7FF4E/data/Containers/Bundle/Application/CC8B9CFC-6B25-4244-872A-C49F874C523A/mocha.app/ti-mocha.js:4290:32</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8d3ca42c5b45b4081f5cdde446bf97d19f4fff9f
</p>

jquick-axway commented on android/titanium/src/java/org/appcelerator/titanium/io/TiFileFactory.java:
```diff
@@ -71,87 +77,94 @@ public static TiBaseFile createTitaniumFile(String path, boolean stream)
+				// Strip this fake base path
+				path = path.substring(ANDROID_ASSET_RESOURCES.length());
+				return new TiResourceFile(trimFront(path, '/')); // remove leading '/' characters
+			}
```
The "file:///android_asset/" URL is not a fake path. It's officially supported by Google and is used to access files from the APK's assets folder. But I'm pretty sure it cannot be loaded by a Java `File` class. This seems to be an issue with the old code. We need to do the below else-if block to turn the given URL into a "content://" so that it can be loaded by our own content provider. Note that I haven't tested this yet. I need to confirm that we can access an APK's root "assets" file such as "file:///android_asset/app.json".
```java
} else if (path.startsWith("/android_asset/")) {
	Uri contentUri = TiFileProvider.createUriFrom(possibleURI).toString();
	if (contentUri != null) {
		return new TitaniumBlob(contentUri.toString());
	}
	return null;
}
```

FYI: The "file:///android_asset/Resources" path needs special handling in Titanium because the given path may reference encrypted assets such as a `.js` or `.json` file. This is where a Titanium project's "Resources" files gets dumped.

jquick-axway replied:
Oh bummer. The above code causes infinite recursion. I'll have to come up with a different solution.

jquick-axway replied:
Ideally, we should wrap a path such as "file:///android_asset/app.json" with our `TiResourceFile` since it supports fast directory listing support (we can't get a directory listing from a "content://" URL which is what my above code is turning it into). Unfortunately, our `TiResourceFile` class is currently limited to files under the "file:///android_asset/Resources" directory, but I believe @garymathews's upcoming snapshot PR will change this since it now wants the "Resources/" base directory.

So, this should be a task for later.

For now, either return `null` for this case or I guess continue to return a broken `TiFile` like the old code was doing before.

jquick-axway commented:
We should grab the relevant file url/path tests from the following PRs and turn them into unit tests. They exercise several different URL schemes and our relative path handling.
- https://github.com/appcelerator/titanium_mobile/pull/10881
- https://github.com/appcelerator/titanium_mobile/pull/10304

Would you like me to help write up the unit tests?

I definitely need to double check our "file:///android_asset/" handling outside of our "Resources" directory too.

sgtcoolguy commented:
@jquick-axway Sorry, I've been bouncing all over the place so didn't see these comments until now. Yeah feel free to hack this up, add tests, or fix it in any way.

These PRs grew out of implementing the Node-compatible `fs` stuff, so as long as they fix the underlying issues I'm ok if you tweak it to be more compatible with other APIs/future anticipated work here.

sgtcoolguy committed:
test(file): add tests around getFile with various URIs

tweak to handle ios/android, mark where parity falls apart

sgtcoolguy commented:
So I've updated this with a number of tests. The reality is **not** pretty. I added a test for the very specific issue I had fixed that was breaking the Node `fs` shim PR (#10998): basically moving/copying with `file:` URIs that din't have the `//` after the colon would break. The test now passes and that's awesome. I'd like to keep this ticket/PR limited to that issue if we can.

I had also tried to add tests around various possible URIs to refer to files, but iOS and Android are woefully different, breaking in different ways. Also, the platforms are inconsistent *internally* in that different APIs handle URIs/paths differently (see above that `Ti.Filesystem.getFile()` doesn't actually surface the fixed bug, but `Ti.Filesystem.File.move()` and `Ti.Filesystem.File.copy()` do!).
I think this is a much larger issue and should likely be dealt with in a separate ticket that tries to enforce parity in URI handling across platforms for `Ti.Filesystem.getFile()`  -and maybe also follow-on tickets to ensure the other APIs play by the same rules too (i.e. Webview references, `Ti.App.appURLToPath()`, `Ti.Platform.canOpenURL()`, `Ti.Database.open/install`, `Ti.Network.HTTPClient.file`, `Ti.Filesystem.File.move()` and `Ti.Filesystem.File.copy()`, `TiUrl` class in Android)

jquick-axway commented:
Personally, I wish our `File` object only supported file system paths; not URLs. But what's done is done. Changing it would probably be too huge of a breaking-change.

I will say that Android "content://" URLs are very commonly used for accessing sandboxed files like photos/videos and they might not actually reference files on the file system (they can be in-memory blobs, database blobs, etc.). In this case, there is no notion of "directory". This also means there is no means of getting a directory listing of files from a "content://". This is by design since it would break the sandboxing of files. (Google does support a database row cursoring feature via `CURSOR_DIR_BASE_TYPE`, but most apps don't implement it in their `ContentProvider`.)

One thing that bothers me is that if we receive a relative path without a scheme, it defaults to the `applicationDataDirectory` on Android. I would expect it to default to `resourcesDirectory`. I haven't had time to check what iOS' default behavior is.
